src/Utilities/TrackedDependencies/CanonicalTrackedInputFiles.cs

```diff
@@ -1071,9 +1073,15 @@ private void RemoveDependenciesFromEntryIfMissing(string rootingMarker)
                     if (keyIndex++ > 0)
                     {
                         // If we are ignoring missing files, then only record those that exist
-                        if (FileUtilities.FileExistsNoThrow(file))
+                        // Cache the files as we find them to save time (On^2), at the expense of storing data O(n).
+                        if (fileCache.Contains(file) || FileUtilities.FileExistsNoThrow(file))
```

Comment:
The only reason I see not to have it on is that the values could become stale. This captures only half of that, so if a file is added after it was first determined to not exist, this would still be able to pick it up, whereas the other cache would not. It sounds better to go with the other cache to me, but since it's behind an escape hatch, I'm guessing someone actually did that at some point? Do we know if they added a file or deleted a file and whether their case generalizes?

Reply:
I actually didn't catch this, whoops! I'm not sure why it's behind an escape hatch, but It's worth investigating

Nathan raises a good question. I had the same [question](https://github.com/microsoft/msbuild/issues/5180#issuecomment-634994143) that I forget the answer to (cc: @rainersigwald ). What should be done about stale values? Does it even matter here?

Reply:
That escape hatch is for a different thing.

In general, it's not safe to cache file existence checks in MSBuild: we're a build engine! The whole point is to create files! So knowing that a file didn't exist N milliseconds ago is not generally useful: was that just the "should I run this target?" check, and now we're doing copy-if-exists after creating it?

The escape hatch was added for fast-evaluation scenarios--if you pinkie-swear to not do any operations that could create files as part of the "build"/evaluation you're doing, you can avoid many checks that would be required for correctness without that constraint. The internal CloudBuild system uses that to quickly use heuristics to predict project inputs and outputs.

Reply:
If this also used for Clean? For a simple build, you wouldn't be removing a file unless you replace it with a more up-to-date version of itself, and that would make this cache perfect, but if Clean is also relevant, as it probably is, I don't think we can cache anything.

Reply:
> For a simple build, you wouldn't be removing a file unless you replace it with a more up-to-date version of itself, and that would make this cache perfect. 

I don't think that's true in practice. There are plenty of remove/delete operations in a normal build, especially in the face of errors.